### PR TITLE
Solaris and Illumos fixups

### PR DESCRIPTION
**Pull request by dcbaker**
This is just a bit of work I did on top of the dynamic linker split work to get illumos working. There are are still failing tests, but this gets us a little closer and canonicalizes the name to use in `*_machine`

### CODE REVIEW

File: run_unittests.py [hunk 1]
```diff
@@ -5058,21 +5062,24 @@ def test_cross_find_program(self):
...
+
+            [properties]
+
+            [host_machine]
+            system = 'linux'
+            cpu_family = 'arm'
+            cpu = 'armv7' # Not sure if correct.
+            endian = 'little'
+            ''').format(
+                os.path.join(testdir, 'some_cross_tool.py'),
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__ [E126] continuation line over-indented for hanging indent ([link](https://sider.review/gh/repos/19784232/pulls/5820#issue-4458745))

<sub>posted by Sider</sub>

**Comment by jpakkane** [comment]
Apart from the indent issue looks ok. But for the naming issue we should ask @alanc and maybe also someone from the Illumos camp?

**Comment by dcbaker** [comment]
@Mno-hime and @chincheta0815 have both submitted patches and/or issues for illumos. Maybe one of them has an opinion?

File: docs/markdown/Reference-tables.md [hunk 2]
```diff
@@ -96,6 +96,7 @@ These are provided by the `.system()` method call.
 | linux               | |
 | netbsd              | |
 | windows             | Any version of Windows |
+| sunos               | Illumos and Solaris |
```
Comment (Mno-hime) [inline docs/markdown/Reference-tables.md]: Make it lowercase illumos, please.

**Comment by chincheta0815** [comment]
> 
> 
> @Mno-hime and @chincheta0815 have both submitted patches and/or issues for illumos. Maybe one of them has an opinion?

Well, actually I am a little bit away from this topic. Maybe this helps (https://github.com/chincheta0815/meson/tree/add.SunOS) as I got a lot of tests working so far, but there were some issues left.
Concerning the naming I also would say "Solaris ld" and "sunos".

**Comment by Mno-hime** [comment]
@alarcher @citrus-it FYI since you merged meson to your OpenIndiana and, respectively, OmniOS CE illumos distributions, thought you might be interested as well.

**Comment by jpakkane** [comment]
```
LINK : warning LNK4044: unrecognized option '/undefined'; ignored
LINK : warning LNK4044: unrecognized option '/headerpad_max_install_names'; ignored
LINK : fatal error LNK1181: cannot open input file 'dynamic_lookup.obj'
```

Errr? Needs a rebase? Or something else?

**Comment by jpakkane** [comment]
The linker changes are in. So a rebase should (have a possibility to) make this work?

**Comment by dcbaker** [comment]
@jpakkane, I haven't retested on illumos, but CI is green.

**Comment by dcbaker** [comment]
Tested on illumos. There's still a few failures, but things look pretty good. As good as the BSDs at least.

**Comment by jpakkane** [comment]
Objective C fails on Clang:

`´´
clang @prog.exe.rsp
prog@exe\prog.m.obj : fatal error LNK1112: module machine type 'x64' conflicts with target machine type 'x86'
clang.exe: error: linker command failed with exit code 1112 (use -v to see invocation)
```

**Comment by alanc** [comment]
I have some of these same fixes in my local tree that I hadn't pushed to github yet - so I'm happy to see them upstream:
- environment: gcc/g++ and clang/clang++ are valid objc/objc++ compilers
- tests: add sunos to get define 

I'd taken the easy way out on the cross file tests and just disabled them - the fix to use gcc instead of /usr/bin/cc looks like a better solution.

I hadn't yet hit the problem in test 162, but the analysis there is correct - cp/mv/ln on Solaris/illumos are a single binary, hardlinked to all three paths and relying on argv[0] to determine behavior.

I need to rebase my test patches from 0.51 to git master, and will try to get them submitted separately soon, but thank @dcbaker for this work and look forward to seeing it merge.

**Comment by jpakkane** [comment]
I'm going to be less online than usual in the next two week, so if you can get the issue fixed, feel free to merge.

**Pull request merged**